src/charm.py

```diff
@@ -0,0 +1,168 @@
+        self.framework.observe(self.on.get_jdbc_endpoint_action, self._on_get_jdbc_endpoint)
+
+    def _on_install(self, event: ops.InstallEvent) -> None:
+        """Handle the `on_install` event."""
+        self.unit.status = Status.WAITING_PEBBLE.value
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
+        """Handle the on_config_changed event."""
+        if not self.unit.is_leader():
+            return
+
+        self.update_service()
+
+    def _update_event(self, _):
+        """Handle the update event hook."""
+        self.unit.status = self.get_status()
+
+    def _update_spark_configs(self):
+        """Update Spark properties in the spark-defaults file inside the charm container."""
+        s3_info = self.s3_connection_info
+        namespace = self.config[NAMESPACE_CONFIG_NAME]
+        service_account = self.config[SERVICE_ACCOUNT_CONFIG_NAME]
+        with self.workload.get_spark_configuration_file(IOMode.WRITE) as fid:
+            spark_config = KyuubiServerConfig(
+                s3_info=s3_info, namespace=namespace, service_account=service_account
+            )
+            fid.write(spark_config.contents)
+
+    def get_status(
+        self,
+    ) -> ops.StatusBase:
+        """Compute and return the status of the charm."""
+        if not self.workload.ready():
+            return Status.WAITING_PEBBLE.value
+
+        s3_info = self.s3_connection_info
+        if not s3_info:
+            return Status.MISSING_S3_RELATION.value
+
+        if not s3_info.verify():
+            return Status.INVALID_CREDENTIALS.value
+
+        namespace = self.config[NAMESPACE_CONFIG_NAME]
+        if not k8s_utils.is_valid_namespace(namespace=namespace):
+            return Status.INVALID_NAMESPACE.value
+
+        service_account = self.config[SERVICE_ACCOUNT_CONFIG_NAME]
+        if not k8s_utils.is_valid_service_account(
+            namespace=namespace, service_account=service_account
+        ):
+            return Status.INVALID_SERVICE_ACCOUNT.value
+
+        return Status.ACTIVE.value
+
+    def update_service(
+        self,
+    ) -> bool:
+        """Update the Kyuubi server service if needed."""
+        # Set the unit status
+        status = self.get_status()
+        self.unit.status = status
+
+        if status is not Status.ACTIVE.value:
+            logger.info(f"Cannot start service because of status {status}")
+            self.workload.stop()
+            return False
+
+        # Dynamically update the Spark properties
+        self._update_spark_configs()
+
+        # Start the workload
+        self.workload.start()
+        return True
+
+    def _on_kyuubi_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Define and start a workload using the Pebble API."""
+        logger.info("Kyuubi pebble service is ready.")
+        self.update_service()
+
+    def _on_get_jdbc_endpoint(self, event: ActionEvent):
```

Comment:
**out-of-scope** this will probably be replaced by some relations (e.g. data-integrator)

Reply:
That is noted. The action is a temporary implementation so we get the first working implementation of the charm ready.